### PR TITLE
feat(core): make DB more stable by checking OS file and map limits on startup and SELECTs

### PR DESCRIPTION
**Pull request by ideoma**
Inspired by https://github.com/questdb/rfc/discussions/63

On Linux check `vm.max_map_count` and open file limit on process startup. If below configured counts fail to start

```
# Switch to enable / disable checks against OS parameters of vm.max_map_count and open file limit
#cairo.check.os.process.limits=true

# Minimum process file descriptor limit to check against at startup
#cairo.check.os.process.files.limit=100000

# Minimum vm.max_map_count limit for the process to check against at startup
#cairo.check.os.process.maps.limit=262144
```
To start QuestDB without checks environment variable `QDB_CHECK_OS_PROCESS_LIMITS` has to be set to false.

Also on every SELECT query, it is estimated of how many column files has to be opened / mapped. If the estimated file count is below available capacity, the query will be failed.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 128 / 155 (82.58%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|1|3|33.33%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|6|12|50.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|18|29|62.07%|
|:large_blue_circle:|io/questdb/std/OsUtils.java|28|34|82.35%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|13|15|86.67%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|45|45|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractDataFrameCursorFactory.java|3|3|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 128 / 155 (82.58%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|1|3|33.33%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|6|12|50.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|18|29|62.07%|
|:large_blue_circle:|io/questdb/std/OsUtils.java|28|34|82.35%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|13|15|86.67%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|45|45|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractDataFrameCursorFactory.java|3|3|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 128 / 155 (82.58%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|1|3|33.33%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|6|12|50.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|18|29|62.07%|
|:large_blue_circle:|io/questdb/std/OsUtils.java|28|34|82.35%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|13|15|86.67%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|45|45|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractDataFrameCursorFactory.java|3|3|100.00%|

File: core/src/main/java/io/questdb/PropServerConfiguration.java [hunk 1]
```diff
@@ -430,6 +433,9 @@ public PropServerConfiguration(
         this.snapshotInstanceId = getString(properties, env, PropertyKey.CAIRO_SNAPSHOT_INSTANCE_ID, "");
         this.snapshotRecoveryEnabled = getBoolean(properties, env, PropertyKey.CAIRO_SNAPSHOT_RECOVERY_ENABLED, true);
         this.simulateCrashEnabled = getBoolean(properties, env, PropertyKey.CAIRO_SIMULATE_CRASH_ENABLED, false);
+        this.checkOsProcessLimits = getBoolean(properties, env, PropertyKey.CAIRO_CHECK_OS_PROCESS_LIMITS, true);
+        this.checkOsProcessLimitFiles = getLong(properties, env, PropertyKey.CAIRO_CHECK_OS_PROCESS_FILES_LIMIT, 100_000);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/PropServerConfiguration.java:437]: naming: `OS` is redundant

File: core/src/main/java/io/questdb/PropServerConfiguration.java [hunk 2]
```diff
@@ -430,6 +433,9 @@ public PropServerConfiguration(
...
+        this.checkOsProcessLimitMaps = getLong(properties, env, PropertyKey.CAIRO_CHECK_OS_PROCESS_MAPS_LIMIT, 262_144);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/PropServerConfiguration.java:438]: `processMinMaps` pls rename constant, field and getter

File: core/src/main/java/io/questdb/PropServerConfiguration.java (hunk 1 above)
Comment (bluestreak01) [inline core/src/main/java/io/questdb/PropServerConfiguration.java:437]: `processMinFiles`

File: core/src/main/java/io/questdb/std/OsUtils.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import io.questdb.std.str.StringSink;
+
+public class OsUtils {
+    public static long getMaxMapCount(Log log, FilesFacade ff) {
+        if (Os.type == Os.LINUX_AMD64 || Os.type == Os.LINUX_ARM64) {
+            return getMaxMapCountLinux(log, ff);
+        }
+
+        if (Os.type == Os.OSX_ARM64 || Os.type == Os.OSX_AMD64) {
+            return getMaxMapCountOsx(log, ff);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/std/OsUtils.java:38]: unused args

File: core/src/main/java/io/questdb/std/OsUtils.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+ ******************************************************************************/
+
+package io.questdb.std;
+
+import io.questdb.log.Log;
+import io.questdb.std.str.Path;
+import io.questdb.std.str.StringSink;
+
+public class OsUtils {
+    public static long getMaxMapCount(Log log, FilesFacade ff) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/std/OsUtils.java:32]: nit: may be have this method on FilesFacade?

File: core/src/main/java/io/questdb/std/OsUtils.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        if (Os.type == Os.LINUX_AMD64 || Os.type == Os.LINUX_ARM64) {
+            return getMaxMapCountLinux(log, ff);
+        }
+
+        if (Os.type == Os.OSX_ARM64 || Os.type == Os.OSX_AMD64) {
+            return getMaxMapCountOsx(log, ff);
+        }
+
+        return -1;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/std/OsUtils.java:41]: windows file limit seems to be 16*1024*1024

File: core/src/main/java/io/questdb/cairo/AbstractDataFrameCursorFactory.java [hunk 6]
```diff
@@ -47,12 +47,14 @@ public void toSink(CharSink sink) {
     }
 
     protected TableReader getReader(CairoSecurityContext sqlContext) {
-        return engine.getReader(
+        TableReader reader = engine.getReader(
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/AbstractDataFrameCursorFactory.java:50]: there are 139 instances where getReader() is used, the check should be done at pool level

File: core/src/main/java/io/questdb/cairo/TableReader.java [hunk 7]
```diff
@@ -141,6 +143,41 @@ public static int getPrimaryColumnIndex(int base, int index) {
         return 2 + base + index * 2;
     }
 
+    /**
+     * Estimate if it's possible to do open all the files using this reader.
+     * Check that there is still enough of Open File Limits and Map limits left to open all partitions.
+     * If capacity is estimated to be not enough, CairoException will the thrown.
+     */
+    public void checkOsLimitsCapacity() {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/TableReader.java:151]: can we maintain count of non-zero partitions so that we don't have to iterate this list?

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/questdb/cairo/AbstractDataFrameCursorFactory.java (hunk 6 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/AbstractDataFrameCursorFactory.java:50]: That would be too strict I believe and make failures in places where Reader is taken to compile a query or to read symbols in ILP.
This is the point where we are about to execute SELECT and checking capacity here will fail the read query

File: core/src/main/java/io/questdb/std/OsUtils.java (hunk 4 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/std/OsUtils.java:32]: It's useful to have some logging facility here, FilesFacade currently does not have Log dependency (neither as a member, nor as method parameters)

File: core/src/main/java/io/questdb/PropServerConfiguration.java (hunk 1 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/PropServerConfiguration.java]: done in https://github.com/questdb/questdb/pull/2289/commits/5bfdd906251b7ede8d75aa234a5ccfff1892dcec

Comment (ideoma) [reply core/src/main/java/io/questdb/PropServerConfiguration.java]: done in https://github.com/questdb/questdb/pull/2289/commits/5bfdd906251b7ede8d75aa234a5ccfff1892dcec

File: core/src/main/java/io/questdb/cairo/TableReader.java (hunk 7 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/TableReader.java:153]: done in https://github.com/questdb/questdb/pull/2289/commits/945b687d19f7d913e2abb6ccb33a663638ecdb5c

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 175 / 202 (86.63%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|1|3|33.33%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|6|12|50.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|24|35|68.57%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|13|17|76.47%|
|:large_blue_circle:|io/questdb/std/OsUtils.java|28|31|90.32%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|13|14|92.86%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|60|60|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractDataFrameCursorFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/AbstractPool.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 175 / 202 (86.63%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|1|3|33.33%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|6|12|50.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|24|35|68.57%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|13|17|76.47%|
|:large_blue_circle:|io/questdb/std/OsUtils.java|28|31|90.32%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|13|14|92.86%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|60|60|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractDataFrameCursorFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/AbstractPool.java|2|2|100.00%|

**Pull request closed without merging**